src/boss/worker.go

```diff
@@ -2,214 +2,479 @@ package boss
+		// clean workers if target is smaller than current cluster size - cleaning worker - destroying worker
+		// ex) if target = 3, and starting, running, cleaning, and destroying  = 1, 2, 1, 1 respectively
+		//     then, scaleSize = 3 - 5 = -2. toBeClean = -1 since 2 workers in cleaning and destroying will eventually be destroyed.
+		//
+		// ex) if target = 1, and 1, 1, 0, 0
+		//     originally we will shut down the running worker, but this will lead a period of time when no worker is available
+		//     so we substract the starting workers also
+		//     in this case, the program will not shut down workers until the starting worker changes to running status
+		toBeClean := -1*scaleSize - len(pool.workers[CLEANING]) - len(pool.workers[DESTROYING]) - len(pool.workers[STARTING])
+		for i := 0; i < toBeClean; i++ { //TODO: policy: clean worker with least tasks
+			worker := <-pool.queue
+			pool.cleanWorker(worker)
+		}
 	}
-	authorized := strings.Split(string(tmp), "\n")
 
-	matches := false
-	for _, v := range authorized {
-		if strings.TrimSpace(v) == pub {
-			matches = true
+	// recover workers if target - starting worker - running worker > 0
+	// ex) if target = 5, and running, cleaning, and destroying  = 1, 2, 1, 1 respectively
+	//     then, toBeRecover = 5 - 1 - 2 = 2 and recover 1 cleaning worker since destroying worker cannot be recovered
+	toBeRecover := pool.target - len(pool.workers[STARTING]) - len(pool.workers[RUNNING])
+	for _, worker := range pool.workers[CLEANING] {
+		if toBeRecover <= 0 { //TODO: policy: recover worker with most tasks
 			break
 		}
+		pool.recoverWorker(worker)
+		toBeRecover--
 	}
+}
 
-	if !matches {
-		return nil, fmt.Errorf("could not find id_rsa.pub in authorized_keys, consider running: cat ~/.ssh/id_rsa.pub >> ~/.ssh/authorized_keys ")
+//run lambda function
+func (pool *WorkerPool) RunLambda(w http.ResponseWriter, r *http.Request) {
+	starttime := time.Now()
+	if len(pool.workers[STARTING])+len(pool.workers[RUNNING]) == 0 {
+		w.WriteHeader(http.StatusInternalServerError)
+		if Conf.Scaling == "manual" {
+			_, err := w.Write([]byte("no active worker\n"))
+			if err != nil {
+				log.Printf("no active worker: %s\n", err.Error())
+			}
+			return
+		}
 	}
 
-	fmt.Printf("STEP 1: get access token\n")
-	client, err := NewGCPClient("key.json")
-	if err != nil {
-		return nil, err
-	}
+	worker := <-pool.queue
+	pool.queue <- worker
 
-	fmt.Printf("STEP 2: lookup instance from IP address\n")
-	instance, err := client.GcpInstanceName()
-	if err != nil {
-		return nil, err
+	atomic.AddInt32(&worker.numTask, 1)
+	atomic.AddInt32(&totalTask, 1)
+	if Conf.Scaling == "auto" {
+		pool.Scale(pool)
 	}
-	fmt.Printf("Instance: %s\n", instance)
 
-	fmt.Printf("STEP 3: take crash-consistent snapshot of instance\n")
-	disk := instance // assume GCP disk name is same as instance name
-	resp, err := client.Wait(client.GcpSnapshot(disk))
-	fmt.Println(resp)
-	if err != nil {
-		return nil, err
+	if Conf.Platform == "mock" {
```

Comment:
it will be nice not to have special cases for mock very often in the code.  One way you could do this: instead of forwardTask taking a workerIp, it could take the whole worker.  The worker could specify both its IP and port.  We could have the mock instances listening on different ports and actually responding to requests.